File: src/python/problems/max_level_sum_test.py
# https://leetcode.com/problems/maximum-level-sum-of-a-binary-tree

from collections import deque
from typing import Optional
from core.problem_base import ProblemBase, TestGen
from models.tree_node import TreeNode

class MaxLevelSum(ProblemBase):
    def Solution(self, root: Optional[TreeNode]) -> int:
        queue = deque()
        queue.append((root, 1))

        levels = {}
        while queue:
            node, level = queue.popleft()
            levels[level] = levels.get(level, 0) + node.val

            if node.left:                
                queue.append((node.left, level + 1))

            if node.right:
                queue.append((node.right, level + 1))
        
        result = 1
        max = root.val
        for level, sum in levels.items():
            if sum > max:
                max = sum
                result = level

        return result

if __name__ == '__main__':
    node1 = TreeNode(1, TreeNode(7, TreeNode(7), TreeNode(-8)), TreeNode(0))

    TestGen(MaxLevelSum) \
        .Add(lambda tc: tc.Param(node1).Result(2)) \
        .Run()

File: src/go/challenges/problems/max_level_sum_test.go
// https://leetcode.com/problems/maximum-level-sum-of-a-binary-tree/

package problems

import (
	"testing"
)

type MaxLevelSum struct {}

func TestMaxLevelSum(t *testing.T) {
	var node1 = TreeNode{
        Val: 1, 
        Left: &TreeNode{Val: 7, Left: &TreeNode{Val: 7}, Right: &TreeNode{Val: -8}}, 
        Right: &TreeNode{Val: 0}}
	
	result := MaxLevelSum{}.Solution(&node1)
	t.Log(result)
}

func (MaxLevelSum) Solution(root *TreeNode) int {
    queue := []*TreeNode{root}

    result := 1
    maxSum := root.Val
    level := 1
    for len(queue) > 0 {
        sum := 0
        length := len(queue)

        for i:= 0; i < length; i++ {
            node := queue[0]
            queue = queue[1:]

            sum += node.Val

            if node.Left != nil {
                queue = append(queue, node.Left)
            }

            if node.Right != nil {
                queue = append(queue, node.Right)
            }
        }

        if sum > maxSum {
            maxSum = sum
            result = level
        }

        level++
    }

    return result
}